tools/opam/gen_ocaml_config.ml

```diff
@@ -0,0 +1,161 @@
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* For as long it remains not totally impractical to do so, this script is
+   written in OCaml 3.07. Its purpose is to generate an opam .config file
+   containing the following variables:
+   - native: True if ocamlopt is located with ocaml
+   - native-tools: True if ocamlc.opt is located with ocaml
+   - native-dynlink: True if dynlink.cmxa exists in -I + or -I +dynlink
+   - stubsdir: Content of +ld.conf in CAML_LD_LIBRARY_PATH format
+   - preinstalled: True if this installation is provided by the system, rather
+                   than compiled from sources by opam
+   - compiler: ["system"], if [preinstalled], otherwise the version of the opam
+               compiler package which provided the compiler (e.g. "5.4.0"). For
+               largely historical reasons, custom compiler append additional
+               configuration information (e.g. "5.4.0+options+flambda"). This
+               variable should be considered deprecated and the content
+               unstable. *)
+
+(* The script must be invoked using the interpreter, for example:
+      ocaml gen_ocaml_config.ml 5.4.0 ocaml 5.4.0+options false +flambda
+   where "5.4.0" is the expected value of Sys.ocaml_version (with any additional
+   information removed), the resulting configuration should be written to
+   "ocaml.config" and the "compiler" variable should be set to
+   "5.4.0+options+flambda". *)
+let expected_ocaml_version,
+    package_config_file,
+    compiler_package_version,
+    preinstalled,
+    option_names =
+  match Array.to_list Sys.argv with
+  | _ ::
+    expected_ocaml_version ::
+    package_config_file ::
+    compiler_package_version ::
+    preinstalled ::
+    options ->
+      expected_ocaml_version,
+      package_config_file ^ ".config",
+      compiler_package_version,
+      preinstalled,
+      String.concat "" (List.filter ((<>) "") options)
+  | _ ->
+      prerr_endline "Invalid arguments";
+      exit 1
+
+(* Check that Sys.ocaml_version is as expected *)
+let () =
+  let ocaml_version =
+    Scanf.sscanf Sys.ocaml_version "%u.%u" (fun major minor ->
+      if (major, minor) > (3, 7) then
+        (* Strip off any additional information *)
+        Scanf.sscanf Sys.ocaml_version "%[^~+]" (fun x -> x)
+      else
+        Sys.ocaml_version)
+  in
+  if ocaml_version <> expected_ocaml_version then
+    (Printf.eprintf
+       "OCaml version mismatch: %s, expected %s"
+       ocaml_version expected_ocaml_version;
+     exit 1)
```

Comment:
As mentioned in https://github.com/ocaml/opam-repository/pull/28455, i don't think we should keep this check

Reply:
This is a different script from that one. The check here is ensuring that, say, the ocaml.5.4.0 package is always installed  with a compiler reporting version 5.4.0: here it's a sanity check that I think we should keep. I haven't forgotten about pulling ocaml-system's version of this script into this PR.